apps/root/src/pages/token-profile/historical-prices/index.tsx

```diff
@@ -0,0 +1,356 @@
+              <TokenIcon token={value.token} size={5} />
+              <Typography variant="bodySmallBold" color={({ palette: { mode } }) => colors[mode].typography.typo2}>
+                {formatCurrencyAmount({ amount: value.amount.amount, token: value.token, intl })}
+              </Typography>
+              <Typography variant="bodySmallRegular" color={({ palette: { mode } }) => colors[mode].typography.typo3}>
+                (${formatUsdAmount({ intl, amount: value.amount.amountInUSD })})
+              </Typography>
+            </ContainerBox>
+          </ContainerBox>
+        ))
+      )}
+    </StyledTooltipContainer>
+  );
+};
+
+const PeriodMap: Record<Exclude<GraphContainerPeriods, GraphContainerPeriods.all>, TimeString> = {
+  [GraphContainerPeriods.day]: '30m',
+  [GraphContainerPeriods.week]: '3h',
+  [GraphContainerPeriods.month]: '12h',
+  [GraphContainerPeriods.year]: '1w',
+};
+
+const DATA_POINTS = 200;
+
+const TokenHistoricalPrices = ({ token }: TokenHistoricalPricesProps) => {
+  const mode = useThemeMode();
+  const priceService = usePriceService();
+  const providerService = useProviderService();
+  const [selectedPeriod, setSelectedPeriod] = React.useState<GraphContainerPeriods>(GraphContainerPeriods.month);
+  const prevPeriod = usePrevious(selectedPeriod);
+  const [isLoadingHistorical, setIsLoadingHistorical] = React.useState(true);
+  const [graphPrices, setGraphPrices] = React.useState<Partial<Record<GraphContainerPeriods, PriceResult[]>>>({});
+
+  const handlePeriodUpdate = (newPeriod: GraphContainerPeriods) => {
+    if (!graphPrices[newPeriod]) {
+      // When period changes, the useMemo renders no data before the isLoading is set to true by the useEffect
+      // By setting it here, we avoid the flickering of the graph
+      setIsLoadingHistorical(true);
+    }
+
+    setSelectedPeriod(newPeriod);
+  };
+
+  React.useEffect(() => {
+    const handleAllTimeframe = async () => {
+      const firstTimestamp = await providerService.getBlockTimestamp(token.chainId, 1n);
+      const currentTimestamp = Math.floor(Date.now() / 1000);
+
+      const totalDataPoints = Math.floor((currentTimestamp - firstTimestamp) / (24 * 60 * 60));
+
+      if (totalDataPoints < 30) {
+        // Fresh chain
+        return {
+          period: PeriodMap[GraphContainerPeriods.day],
+          span: DATA_POINTS,
+        };
+      }
+
+      let span = totalDataPoints;
+
+      let periodDays = 1;
+      while (span > DATA_POINTS) {
+        span = Math.floor(span / 3);
+        periodDays = periodDays * 3;
+      }
+
+      return {
+        period: `${periodDays}d` as TimeString,
+        span,
+      };
+    };
+
+    const fetchGraphPrices = async () => {
+      const priceResult = await priceService.getPricesForTokenGraph({
+        token,
+        ...(selectedPeriod === GraphContainerPeriods.all
+          ? await handleAllTimeframe()
+          : { period: PeriodMap[selectedPeriod], span: DATA_POINTS }),
```

Comment:
Should we actually define how many data points we want based on the selected period?

Reply:
And btw, we could extend longer periods with the data of smaller periods if we have already loaded them

Reply:
I spent some time trying to find the 'universal query' for this graph, but didn't get to any result in that logic. I have refined a bit how I construct the periods based on a span of `200`, keeping this examples in mind:
- For `1d` as selected period, I'm showing data within 24 hours (period: 10m, span: 200 points => 2000 minutes => 30 hours)
- For `1y` as selected period, I'm showing data in a whole year (period: 2d, span: 200 points => 400 days => 1,09 years) 

I think SOME points can be recycled, but the graph would need to refetch some specific data in order to render a more precise chart

Reply:
It makes a bit more sense, I think we should still render more points when its a more longer period, having the daily selected as default would allow us to keep the data more consistent